### PR TITLE
[DDW-823] Detect Internet connection status

### PR DESCRIPTION
**Pull request by yakovkaravelov**
This PR implemented internet connection status overlay
[YouTrack Ticket](https://iohk.myjetbrains.com/youtrack/issue/DDW-823)

## Todos
- [x] Implement internet connection status overlay

## Screenshots
image
image

## Testing Checklist
- [Slack QA thread](https://input-output-rnd.slack.com/archives/GGKFXSKC6/p1566482366254100)
- [x] Watch app when internet is offline to see if `internet offline` alert overlay pops up
- [x] Go to `Internet connection Status` => `Offline - default` story on storybook and check if it matches with wireframe
- [x] Go to `Internet connection Status` => `Offline - with checking spinner` story on storybook and check if it matches with wireframe

### Basics

- [ ] PR has been assigned and has appropriate labels (`feature`/`bug`/`chore`, `release-x.x.x`)
- [ ] PR is updated to the most recent version of the target branch (and there are no conflicts)
- [ ] PR has a good description that summarizes all changes and shows some screenshots or animated GIFs of important UI changes
- [ ] CHANGELOG entry has been added to the top of the appropriate section (*Features*, *Fixes*, *Chores*) and is linked to the correct PR on GitHub
- [ ] Automated tests: All acceptance and unit tests are passing (`yarn test`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *development* build (`yarn dev`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *production* build (`yarn package` / CI builds)
- [ ] There are no *flow* errors or warnings (`yarn flow:test`)
- [ ] There are no *lint* errors or warnings (`yarn lint`)
- [ ] There are no *prettier* errors or warnings (`yarn prettier:check`)
- [ ] There are no missing translations (running `yarn manage:translations` produces no changes)
- [ ] Text changes are proofread and approved (Jane Wild / Amy Reeve)
- [ ] Japanese text changes are proofread and approved (Junko Oda)
- [ ] UI changes look good in all themes (Alexander Rukin)
- [ ] Storybook works and no stories are broken (`yarn storybook`)
- [ ] In case of dependency changes `yarn.lock` file is updated

### Code Quality
- [ ] Important parts of the code are properly commented and documented
- [ ] Code is properly typed with flow
- [ ] React components are split-up enough to avoid unnecessary re-renderings
- [ ] Any code that only works in main process is neatly separated from components

### Testing
- [ ] New feature/change is covered by acceptance tests
- [ ] New feature/change is manually tested and approved by QA team
- [ ] All existing acceptance tests are still up-to-date
- [ ] New feature/change is covered by Daedalus Testing scenario
- [ ] All existing Daedalus Testing scenarios are still up-to-date

### After Review
- [ ] Merge the PR
- [ ] Delete the source branch
- [ ] Move the ticket to `done` column on the YouTrack board
- [ ] Update Slack QA thread by marking it with a green checkmark

### CODE REVIEW

**Comment by gnpf** [comment]
@yakovkaravelov The logo is not visible on the white theme.

image


Also, I'm not sure if this is the expected color for the logo in the yellow theme.
@a-rukin could you confirm this, please?

image

**Comment by gnpf** [comment]
@yakovkaravelov Sometimes the screen is displayed on the default theme for a second before switching to the current theme:

u0iGtmNRlq

**Comment by a-rukin** [comment]
@gnpf @yakovkaravelov image of "daedalus disconnetced from a globe" should be in the color of a text presented on the overlay

**Comment by nikolaglumac** [comment]
@yakovkaravelov please always add a screenshot of Japanese version too 🙏

File: source/renderer/app/config/urlsConfig.js [hunk 1]
```diff
@@ -13,6 +13,7 @@ export const MAINNET_LATEST_VERSION_INFO_URL =
 export const TESTNET_LATEST_VERSION_INFO_URL =
   'updates-cardano-testnet.s3.amazonaws.com';
 export const STAGING_LATEST_VERSION_INFO_URL = 'update-awstest.iohkdev.io';
+export const INTERNET_PING_HOSTNAME = 'ipv4.icanhazip.com';
```
Comment (nikolaglumac) [inline source/renderer/app/config/urlsConfig.js]: @yakovkaravelov can you try using this instead?
```
$ curl -i https://www.gstatic.com/generate_204

HTTP/2 204 
date: Mon, 19 Aug 2019 20:31:49 GMT
alt-svc: quic=":443"; ma=2592000; v="46,43,39"
```

Comment (nikolaglumac) [reply source/renderer/app/config/urlsConfig.js]: Of course, we are not expected to use `curl` in Daedalus...

**Comment by nikolaglumac** [comment]
@tomothespian please review this one too 🙏

File: source/renderer/app/config/urlsConfig.js [hunk 2]
```diff
@@ -21,4 +22,5 @@ export const ALLOWED_EXTERNAL_HOSTNAMES = [
   MAINNET_LATEST_VERSION_INFO_URL,
   TESTNET_LATEST_VERSION_INFO_URL,
   STAGING_LATEST_VERSION_INFO_URL,
+  INTERNET_PING_HOSTNAME,
```
[the remark this answers is not in the document]
Comment (yakovkaravelov) [reply source/renderer/app/config/urlsConfig.js]: @tomothespian No, it's not url, it's host name. You know url is a combination of `protocol`, `hostname`, `port`, `path`, and `query`.
And that constant is just hostname.
Inside `externalRequest`, HttpOptions variable also treats it as `hostname`.
So I'd rather go with `hostname` for this variable.

File: source/renderer/app/api/utils/externalRequest.js [hunk 3]
```diff
@@ -14,7 +14,9 @@ export type HttpOptions = {
   },
 };
 
-export const externalRequest = (httpOptions: HttpOptions): Promise<any> =>
+export const externalRequestForRawBody = (
```
[the remark this answers is not in the document]
Comment (yakovkaravelov) [reply source/renderer/app/api/utils/externalRequest.js]: Last week, when you and I called, I was planning to use it same function with additional parameter, but I faced some flow test error after doing it. So I simply moved to another function. But now I think I can resolve that flow test error, so I will make one function with additional parameter.

**Comment by yakovkaravelov** [comment]
> @yakovkaravelov
> Please take a look on code comments I made.
> 
> Also there are few more things:
> 
> 1. Implement Storybook story
> 2. When App is loaded and I loose internet connection, check is not triggered and dialog not show. Check only works if you lost internet connection before starting app
> 3. Theme css variables naming is not descriptive. ` --theme-network-window...` should be something more similar to `theme-offline-dialog...`. network window is not defined here but offline / connection / internet are proper keys so use some and combine into proper variable name.
> 4. Define new variables also in `themes/utils/createTheme` not just in specific theme.

1. OK. I will do it.
2. You couldn't find dialog not shown because you were running Cardano-sl repo on your local machine.
In real world, this daedalus app will be run on different machine than cardano-sl repo, and it will run ok.
3. I used `theme-network-window` key after taking a look at `Network disconnected. Synching now` dialog which was already implemented before this ticket.
That dialog also checks network status(but not internet connection which current ticket is working for), so I followed it. Maybe I should use key like `theme-internet-window`.
4. OK, I will do it.

@tomothespian I will do the remaining things.
But would you please take a look at the comments I posted?
Thanks for the review, @tomothespian

**Comment by nikolaglumac** [comment]
@yakovkaravelov I agree with your comment regarding point #2

File: source/renderer/app/api/utils/externalRequest.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (tomothespian) [reply source/renderer/app/api/utils/externalRequest.js]: @yakovkaravelov ok, do what you think that is best but not on way how is implemented yet.

File: source/renderer/app/config/urlsConfig.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (tomothespian) [reply source/renderer/app/config/urlsConfig.js]: @yakovkaravelov I agree, but use something more descriptive like `CHACK_INTERNET_CONNECTION_HOSTNAME` or similar

**Comment by tomothespian** [comment]
@yakovkaravelov please take a look on my comments and also merge latest `develop` branch

File: source/renderer/app/api/utils/externalRequest.js [hunk 4]
```diff
@@ -33,23 +34,20 @@ export const externalRequestForRawBody = (
         body += chunk;
       });
       response.on('error', error => reject(error));
-      response.on('end', () => resolve(body));
-    });
-    request.on('error', error => reject(error));
-    return request.end();
-  });
+      response.on('end', () => {
+        if (expectRawBody) {
```
Comment (tomothespian) [inline source/renderer/app/api/utils/externalRequest.js]: @yakovkaravelov put this one also under `try`

Comment (tomothespian) [reply source/renderer/app/api/utils/externalRequest.js]: @yakovkaravelov you're in right

**Comment by tomothespian** [comment]
> @yakovkaravelov I can still reproduce this issue:
> 
> o9NQ0d2Bw0
> 
> So far, this only happened in Linux for me.
> 
> I'd like to point out a few more things I noticed. They can be considered small improvements, maybe to be implemented for this or another card:
> 
> 1 - There is a delay when trying to load the elements of the screen, and they end up showing up at different times. Is there a way this can be improved?
> 
> 4L6mpQ7wv8
> 
> 2 - When the user clicks the button "Check again", there is no indication that the application is actually checking and maybe it can take a few seconds. Maybe a spinner on the button could be implemented or the button can be disabled so it's clear that the action was triggered. I don't remember if something similar was applied before.
> 
> Let me know what you think.

@a-rukin Please take a look on point 2. Can add some suggestion?

File: source/renderer/app/stores/NetworkStatusStore.js [hunk 5]
```diff
@@ -329,6 +336,27 @@ export default class NetworkStatusStore extends Store {
 
   // DEFINE ACTIONS
 
+  @action updateInternetConnectionStatus = async () => {
+    try {
+      await externalRequest(
+        {
+          hostname: CHECK_INTERNET_PING_HOSTNAME,
```
Comment (tomothespian) [inline source/renderer/app/stores/NetworkStatusStore.js]: @yakovkaravelov something like `CHECK_INTERNET_CONNECTION_HOSTNAME ` is more descriptive.

File: source/renderer/app/config/urlsConfig.js [hunk 6]
```diff
@@ -13,6 +13,7 @@ export const MAINNET_LATEST_VERSION_INFO_URL =
 export const TESTNET_LATEST_VERSION_INFO_URL =
   'updates-cardano-testnet.s3.amazonaws.com';
 export const STAGING_LATEST_VERSION_INFO_URL = 'update-awstest.iohkdev.io';
+export const CHECK_INTERNET_PING_HOSTNAME = 'ipv4.icanhazip.com';
```
Comment (tomothespian) [inline source/renderer/app/config/urlsConfig.js]: @yakovkaravelov , please use what @nikolaglumac suggested here: https://github.com/input-output-hk/daedalus/pull/1546#discussion_r317097073
`https://www.gstatic.com/generate_204`

File: source/renderer/app/api/utils/externalRequest.js [hunk 7]
```diff
@@ -32,12 +35,16 @@ export const externalRequest = (httpOptions: HttpOptions): Promise<any> =>
...
           const parsedBody = JSON.parse(body);
-          resolve(parsedBody);
+          return resolve(parsedBody);
         } catch (error) {
           // Handle internal server errors (e.g. HTTP 500 - 'Something went wrong')
-          reject(new Error(error));
+          return reject(new Error(error));
         }
       });
     });
```
Comment (tomothespian) [inline source/renderer/app/api/utils/externalRequest.js]: @yakovkaravelov Please check what server you are trying to reach will return if there is some server failure. Throw error and show "No Connection" dialog only if app is disconnected and not if you get server failure (you can have stable internet connection in that case and dialog will appear)

**Comment by a-rukin** [comment]
Yep, the second point is valid
We should add spinner there (spinner color = text color for those buttons) 
Ping me if you need some clarifications about stypes. But we've already implemented same behaviour

**Comment by IuliaDolishniak** [comment]
> @yakovkaravelov Sometimes the screen is displayed on the default theme for a second before switching to the current theme:
> 
> u0iGtmNRlq

Hi @yakovkaravelov the issue @gnpf has described here is still reproducible for me in Linux too.

**Comment by a-rukin** [comment]
https://zpl.io/awOQPx1 - updated with spinner style

**Comment by tomothespian** [comment]
> > @yakovkaravelov Sometimes the screen is displayed on the default theme for a second before switching to the current theme:
> > u0iGtmNRlq
> 
> Hi @yakovkaravelov the issue @gnpf has described here is still reproducible for me in Linux too.

@yakovkaravelov please check what is the reason and try to resolve. That glitch is not acceptable! 🙏 
@IuliaDolishniak @DmitriiGaico I added one more storybook story, please take a look on that

**Comment by tomothespian** [comment]
@yakovkaravelov I re-added WIP label because theme switching glitch must be resolved!

**Comment by IuliaDolishniak** [comment]
@tomothespian glitch was found by @gnpf I have only confirmed that I am able to reproduce it.

**Comment by yakovkaravelov** [comment]
@gnpf Regarding overlay glitches, I changed code to show other modals as app started, and same thing happened to them as well which means this is generic issue across the entire app.
so this issue is not relying to this PR/ticket.
We need to file a new separated ticket for this and solve it.

**Comment by IuliaDolishniak** [comment]
> @yakovkaravelov The style for the text in Japanese in Storybook (2) does not match the style on Daedalus (1)
> 
> image

@yakovkaravelov what about this issue? will it be fixed or this is storybook problem?

**Comment by IuliaDolishniak** [comment]
> @yakovkaravelov please check what is the reason and try to resolve. That glitch is not acceptable! 🙏
@tomothespian please confirm that you have agreed to move the issue with the glitch to another PR?

**Comment by yakovkaravelov** [comment]
> > @yakovkaravelov The style for the text in Japanese in Storybook (2) does not match the style on Daedalus (1)
> > image
> 
> @yakovkaravelov what about this issue? will it be fixed or this is storybook problem?

@IuliaDolishniak Just had a call with Gabriella, and ensured it's a bug of her test environment.
It should work ok.

**Comment by gnpf** [comment]
> @IuliaDolishniak Just had a call with Gabriella, and ensured it's a bug of her test environment.
> It should work ok.

@IuliaDolishniak and I checked this on other OS and browsers and this is displayed correctly.

> @gnpf Regarding overlay glitches, I changed code to show other modals as app started, and same thing happened to them as well which means this is generic issue across the entire app.
> so this issue is not relying to this PR/ticket.
> We need to file a new separated ticket for this and solve it.

@yakovkaravelov and I had a call about this and I created [a card](https://iohk.myjetbrains.com/youtrack/issue/DDW-870) to address this.

**Comment by tomothespian** [comment]
@yakovkaravelov please refactor next part:
- instead of calling externalRequest directly from NetworkStatusStore create checkInternetConnection `request` and define method in `api.js`.  Flow should be same as is for `getLatestAppVersion` request

**Comment by DmitriiGaico** [comment]
> @yakovkaravelov please refactor next part:
> 
> * instead of calling externalRequest directly from NetworkStatusStore create checkInternetConnection `request` and define method in `api.js`.  Flow should be same as is for `getLatestAppVersion` request

Hopefully this helps. I just tried a VM and it worked just fine.
Unfortunately on my bare metal PC which is under Domain(Policy) it works weird. Even connection is UP and Node is syncing i see that message instantly and can't recover.

**Comment by yakovkaravelov** [comment]
@tomothespian @a-rukin Would you please review this PR?
Thanks.

**Comment by a-rukin** [comment]
image
When spinner we shouldn't be able to hover and press this button again

File: source/renderer/app/stores/NetworkStatusStore.js [hunk 8]
```diff
@@ -679,6 +667,17 @@ export default class NetworkStatusStore extends Store {
   };
 
   // DEFINE COMPUTED VALUES
+  @computed get isInternetConnectionChecking(): boolean {
+    return this.checkInternetConnectionRequest.isExecuting;
+  }
+
+  @computed get isInternetConnectionChecked(): boolean {
```
Comment (tomothespian) [inline source/renderer/app/stores/NetworkStatusStore.js]: @yakovkaravelov this one is not necessary

File: source/renderer/app/stores/NetworkStatusStore.js [hunk 9]
```diff
@@ -679,6 +667,17 @@ export default class NetworkStatusStore extends Store {
   };
 
   // DEFINE COMPUTED VALUES
+  @computed get isInternetConnectionChecking(): boolean {
```
Comment (tomothespian) [inline source/renderer/app/stores/NetworkStatusStore.js]: @yakovkaravelov this one is not necessary. Use `checkInternetConnectionRequest.isExecuting` directly in component

File: source/renderer/app/containers/status/InternetConnectionStatusDialog.js [hunk 10]
```diff
@@ -23,8 +23,8 @@ export default class InternetConnectionStatusDialog extends Component<any> {
         ariaHideApp={false}
       >
         <InternetConnectionOfflineStatus
-          checking={checkingInternetConnection}
-          checkAgain={checkInternetConnectionStatus}
+          checking={isInternetConnectionChecking}
```
Comment (tomothespian) [inline source/renderer/app/containers/status/InternetConnectionStatusDialog.js]: @yakovkaravelov use `checkInternetConnectionRequest.isExecuting` and rename property to isCheckingInternetConnectionStatus

**Comment by a-rukin** [comment]
image
Is this button background some solid color instead of rgba(0, 0, 0, 0.1)?

**Comment by a-rukin** [comment]
image
Let's reduce spinner size (but not border-width) to 24px (now 30px), cuz it seems like a bit too close to the button borders

File: source/renderer/app/stores/NetworkStatusStore.js (hunk 9 above)
Comment (tomothespian) [reply source/renderer/app/stores/NetworkStatusStore.js]: @yakovkaravelov but that is not necessary. Why to create some computed value if you have direct access to request flag? 
Please improve that 🙏

File: source/renderer/app/containers/status/InternetConnectionStatusDialog.js (hunk 10 above)
Comment (tomothespian) [reply source/renderer/app/containers/status/InternetConnectionStatusDialog.js]: @yakovkaravelov please use request flag instead computed value. Also don't forget to change property declaration name.

**Comment by yakovkaravelov** [comment]
@tomothespian @a-rukin Please review again.
Thanks.

**Comment by tomothespian** [comment]
> @tomothespian @a-rukin Please review again.
> Thanks.

@yakovkaravelov changes looks good. I will still leave WIP and DO NOT MERGE labels as is because we need to discuss few more things, on Monday, about logic flow and url we are pinging (@DmitriiGaico i got rejection due to firewall settings)

**Comment by tomothespian** [comment]
> > @tomothespian @a-rukin Please review again.
> > Thanks.
> 
> @yakovkaravelov changes looks good. I will still leave WIP and DO NOT MERGE labels as is because we need to discuss few more things, on Monday, about logic flow and url we are pinging (@DmitriiGaico i got rejection due to firewall settings)

@yakovkaravelov please read what I said ⬆️ . Don't move that labels before we discuss that with Nikola!

**Comment by nikolaglumac** [comment]
@yakovkaravelov we are pausing this PR until further notice.

**Comment by nikolaglumac** [comment]
Closing this one as the feature implementation has been abandoned.

**Pull request closed without merging**